README.md

```diff
@@ -87,6 +87,50 @@ The PMEM-CSI driver supports running in different modes, which can be controlled
+down. Data is retained until the volume gets deleted.
+* Ephemeral Volumes
+An application is assigned to a node and when it starts to run, a new volume is created for it on that node. When the application stops, the volume is deleted. This [diagram](/docs/images/sequence/pmem-csi-ephemeral-sequence-diagram.png) illustrates how an ephemeral volume gets provisioned in Kubernetes using PMEM-CSI driver.
+* Cache Volumes
+Volumes are created on a certain set of nodes, each with its own local data. Applications are started on those nodes and then get to use the volume on their node. Data persists across application restarts. This
+is useful when the data is only cached information that can be discarded and reconstructed at any time *and* the application can reuse existing local data when restarting. This [diagram](/docs/images/sequence/pmem-csi-cache-sequence-diagram.png) illustrates how a cache volume gets provisioned in Kubernetes using PMEM-CSI driver.
+
+Volume | Kubernetes | PMEM-CSI | Limitations
+--- | --- | --- | ---
+Persistent | supported | supported | topology aware scheduling
+Ephemeral | [in design](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190122-csi-inline-volumes.md#proposal) | supported | topology aware scheduling, resource constraints
+Cache | supported | supported | topology aware scheduling
+
+[Topology aware scheduling](https://github.com/kubernetes/enhancements/issues/490)
+ensures that an application runs on a node where the volume was created. This is obviously needed for persistent volumes, but also for
+delayed persistent volumes because otherwise a restarted application
+might end up on the wrong node. Usage of cache volumes is more flexible,
+but an application still can only run on one of the nodes where the
+a cache volume is available. For CSI-based drivers like PMEM-CSI,
+Kubernetes >= 1.13 is needed. On older Kubernetes releases, pods must
+be scheduled manually onto the right node(s).
+
+[Topology aware volume provisioning](https://kubernetes.io/blog/2018/10/11/topology-aware-volume-provisioning-in-kubernetes/)
+is needed for delayed persistent volumes.
+
+The upstream design for ephemeral volumes currently does not take
+[resource constraints](https://github.com/kubernetes/enhancements/pull/716#discussion_r250536632)
+into account. If an application gets scheduled onto a node and then
+creating the ephemeral volume on that node fails, the application on
+the node cannot start until resources become available. The same
+applies to how this is currently implemented in PMEM-CSI: a fake
+volume gets created initially, without actually allocating space. Then
+when an application starts and it is time to mount that "existing"
+volume, an attempt is made to allocate the space required for the
+volume locally. Only if there is enough space can application startup
```

Comment:
I thought this section is explaining the volume types in general nothing related to Kubernetes as in your original PR #160. Hence, i provided links to diagrams to illustrate the usage.

I will add "Volumes are created on a certain set of nodes which have free capacity" in case of cache if its not clear for user where those volume are created.

@pohly what additional user documentation you feel useful and which is appropriate section for that.

Reply:
This comment shouldn't have been attached to this particular section. As you said, it is a general introduction to the topic.

What we need is an entirely new "Usage" section that explains what users can put into their `StorageClass` and how the driver will then behave. For a cache, this behavior description must include on how many nodes the cache will be allocated. We probably need some kind of control over that. Imaging a cluster with 100 nodes that have PMEM and an app which will run with a few instances. We don't want to allocate 100 volumes.

Reply:
> We probably need some kind of control over that. Imaging a cluster with 100 nodes that have PMEM and an app which will run with a few instances. We don't want to allocate 100 volumes.

This is not how we operate for cache volumes. Currently the "volume replica/cache count" shall be coming from StorageClass(https://github.com/avalluri/pmem-CSI/blob/046cadf8461d429203db204e46289140d295534f/deploy/kubernetes-1.13/pmem-storageclass-cache.yaml#L10).

@pohly  I would request to review the code.

Reply:
Of course I can reverse-engineer how this new feature works by reading the source. But user-visible documentation needs to be written for it either way, so why not do it now? Then the code review can confirm that the code works as described to the user.